website/docs/cloud-docs/integrations/kubernetes: Update migration guide (#641)
hashicorp/terraform-docs-common: website/docs/cloud-docs/integrations/kubernetes/ops-v2-migration.mdx

$ kubectl get workspace --all-namespaces -o yaml > backup_tfc_operator_v1.yaml
```

## Manifest schema migration

In v2 of the operator, many fields have been renamed or moved. You must update your specification with these changes when you migrate to v2 of the operator.

### Workspace controller

Below is the field mapping of the `Workspace` controller from v1 to v2 of the operator.

> | Version 1 | Version 2 | Notes |

Author: Thanks!

- Renamed the 3rd column to "Changes between versions". It makes more sense.
- Added `None` to the 3rd column where there are no changes between versions and left `N/A` in the 2nd column when a new version does not have a replacement for this field(removed).

| --- | --- | --- |
| `apiVersion: app.terraform.io/v1alpha1` | `apiVersion: app.terraform.io/v1alpha2` | The `apiVersion` is changed from `v1alpha1` to `v1alpha2`. |
| `kind: Workspace` | `kind: Workspace` |  |
| `metadata` | `metadata` |  |
| `spec.organization` | `spec.organization` |  |
| `spec.secretsMountPath` | `spec.token.secretKeyRef` | In v2 the operator keeps HCP Terraform access token in a Kubernetes Secret. |
| `spec.vcs` | `spec.versionControl` | Renamed. Refer to the `versionControl` structure for details. |

> | `spec.vcs.token_id` | `spec.versionControl.oAuthTokenID` | Renamed. Refer to the `versionControl` structure for details. |

Author: Did the same as above: "Renamed the `X` field to `Y`."

| `spec.vcs.repo_identifier` | `spec.versionControl.repository` | Renamed. Refer to the `versionControl` structure for details. |

> | `spec.vcs.branch` | `spec.versionControl.branch` |  |

Author: Added `None` since the end field didn't change. Thanks!

| `spec.vcs.ingress_submodules` | `spec.workingDirectory` | Moved. |
| `spec.variables.[*]` | `spec.environmentVariables.[*]` OR `spec.terraformVariables.[*]` | Variables were split into two different paths. Depending on the variable's value of `spec.variables.environmentVariable` in v1 CRD:<br/> - If it is `true` then you must migrate the variable to `spec.environmentVariables`.<br/> - If it is `false` then you must migrate the variable to `spec.terraformVariables`. |
| `spec.variables.[*]key` | `spec.environmentVariables.[*]name` OR `spec.terraformVariables.[*]name` | The `key` field is renamed to `name`. |

> | `spec.variables.[*]value` | `spec.environmentVariables.[*]value` OR `spec.terraformVariables.[*]value` |  |

Author: I like this idea. Added an empty `a` tag to the explanation row and `Learn more` pointing to it. I hope I get your idea right. :)

| `spec.variables.[*]valueFrom` | `spec.environmentVariables.[*]valueFrom` OR `spec.terraformVariables.[*]valueFrom` |  |
| `spec.variables.[*]hcl` | `spec.environmentVariables.[*]hcl` OR `spec.terraformVariables.[*]hcl` |  |
| `spec.variables.sensitive` |  `spec.environmentVariables.[*]sensitive` OR `spec.terraformVariables.[*]sensitive` |  |
| `spec.variables.environmentVariable` | N/A | Removed, variables are split to `spec.environmentVariables` and `spec.terraformVariables`. |

> | `spec.runTriggers.[*]` | `spec.runTriggers.[*]` |  Changes made to the `runTriggers` structure. |

Author: I agree with you, changed it to `None.`

| `spec.runTriggers.[*].sourceableName` | `spec.runTriggers.[*].name` | The `sourceableName` field is renamed to `name`. |
| `spec.sshKeyID` | `spec.sshKey.id` | The `sshKeyID` is moved to `spec.sshKey.id`. |
| `spec.outputs` | N/A | Removed. |
| `spec.terraformVersion` | `spec.terraformVersion` |  |

> | `spec.notifications` | `spec.notifications` |  Changes made to the `notification` structure. |

Author: Thank you! Agree, changed it to `None.`

<CodeBlockConfig filename="v2">

```yaml
apiVersion: app.terraform.io/v1alpha2
kind: Workspace
metadata:
  name: migration
  spec:
    terraformVariables:
      - name: username
        value: "user"
        hcl: true
        sensitive: false
    environmentVariables:
      - name: SECRET_KEY
        value: "s3cr3t"
        hcl: false
        sensitive: false
```

</CodeBlockConfig>

### Module controller

Terraform Cloud Kubernetes Operator v2 configures modules in a new `Module` controller separate from the `Workspace` controller. Below is a template of a custom resource manifest:

```yaml
---
apiVersion: app.terraform.io/v1alpha2
kind: Module
metadata:
  name: <NAME>
spec:
  organization: <ORG-NAME>
  token:
    secretKeyRef:
      name: <SECRET-NAME>
      key: <KEY-NAME>
  name: operator
```

The table below details the mapping between the `Workspace` controller in v1 of the operator and the `Module` controller in v2 of the operator.

| Version 1 (Workspace CRD) | Version 2 (Module CRD) | Notes |
| --- | --- | --- |
| `spec.module` | N/A | In v2 of the operator a `Module` is a separate controller with its own CRD. |

> | N/A | `spec.name: operator` | In v1 of the operator, the name of the generated module is hardcoded to `operator`. In v2, the default name of the generated module is `this`, however it can be changed. |

Reviewer: ```suggestion
| N/A | `spec.name: operator` | In v1 of the operator, the name of the generated module is hardcoded to `operator`. In v2, the default name of the generated module is `this`, but you can rename it. |
```

| `spec.module.source` | `spec.module.source` | This supports all Terraform [module sources](https://developer.hashicorp.com/terraform/language/modules/sources). |
| `spec.module.version` | `spec.module.version` | Refer to  [module sources](https://developer.hashicorp.com/terraform/language/modules/sources) for versioning information for each module source. |

> | `spec.variables.[*]` | `spec.variables.[*].name` | Variable names should be included in the module. This is a reference to variables in the Workspace where the module will be executed. |

Reviewer: We only capitalize product names:
```suggestion
| `spec.variables.[*]` | `spec.variables.[*].name` | You should include variable names in the module. This is a reference to variables in the workspace that is executing the module. |
```

> | `spec.outputs.[*].key` | `spec.outputs.[*].name` | Output names should be included in the module. This is a reference to the output variables produced by the module. |

Reviewer: ```suggestion
| `spec.outputs.[*].key` | `spec.outputs.[*].name` | You should include output names in the module. This is a reference to the output variables produced by the module. |
```

> | `status.workspaceID` OR `metadata.namespace-metadata.name` | `spec.workspace.id` OR `spec.workspace.name` | The workspace where the module will be executed. The workspace must be within the same organization. |

Reviewer: In the same organization as what? Or are you referring to listing multiple workspace names and those workspaces must be in the same org?
```suggestion
| `status.workspaceID` OR `metadata.namespace-metadata.name` | `spec.workspace.id` OR `spec.workspace.name` | The workspace where the module is executed. The workspace must be in the same organization. |
```

Below is an example migration of a `Module` between v1 and v2 of the operator:

<CodeBlockConfig filename="v1.yaml">

```yaml
apiVersion: app.terraform.io/v1alpha1
kind: Workspace
metadata:
  name: migration
spec:
  module:
    source: app.terraform.io/org-name/module-name/provider
    version: 0.0.42
  variables:
    - key: username
      value: "user"
      hcl: true
      sensitive: false
      environmentVariable: false
    - key: SECRET_KEY
      value: "s3cr3t"
      hcl: false
      sensitive: false
      environmentVariable: true
```

</CodeBlockConfig>

> In v2 of the operator, the workspace and module are manage by separate controllers.

Reviewer: ```suggestion
In v2 of the operator, separate controllers manage workspace and modules.
```

<CodeBlockConfig filename="workspace-v2.yaml">

  ```yaml
  apiVersion: app.terraform.io/v1alpha2
  kind: Workspace
  metadata:
    name: migration
    spec:
      terraformVariables:
        - name: username
          value: "user"
          hcl: true
          sensitive: false
      environmentVariables:
        - name: SECRET_KEY
          value: "s3cr3t"
          hcl: false
          sensitive: false
  ```

</CodeBlockConfig>

<CodeBlockConfig filename="module-v2.yaml">

```yaml
apiVersion: app.terraform.io/v1alpha2
kind: Module
metadata:
  name: migration
spec:
  name: operator
  module:
    source: app.terraform.io/org-name/module-name/provider
    version: 0.0.42
  workspace:
    name: migration
```

</CodeBlockConfig>

## Upgrade the operator

Download Workspace CRD patch A:

```shell-session
$ curl -sO https://raw.githubusercontent.com/hashicorp/terraform-cloud-operator/main/docs/migration/crds/workspaces_patch_a.yaml
```

> View the changes that **patch A** will apply to the workspace CRD.

Reviewer: We only use bold for UI elements:
```suggestion
View the changes that patch A applies to the workspace CRD.
```

```shell-session
$ kubectl diff --filename workspaces_patch_a.yaml
```

> Patch the workspace CRD with **patch A**. This patch adds `app.terraform.io/v1alpha2` support, but excludes `.status.runStatus` because it has a different format in `app.terraform.io/v1alpha1` and causes JSON unmarshalling issues.

Reviewer: ```suggestion
Patch the workspace CRD with patch A. This patch adds `app.terraform.io/v1alpha2` support, but excludes `.status.runStatus` because it has a different format in `app.terraform.io/v1alpha1` and causes JSON un-marshalling issues.
```

> ~> **Warning** Once the patch is applied, Kubernetes will convert existing `app.terraform.io/v1alpha1` custom resources to `app.terraform.io/v1alpha2` according to the updated schema and v1 of the operator will no longer serve custom resources. You must update existing custom resources to satisfy the current schema. Refer to the tables above for a mapping of resource fields between v1 and v2 of the operator.

Reviewer: I think we can shorten this to:
```suggestion
!> **Upgrade warning**: Once you apply a patch, Kubernetes converts existing `app.terraform.io/v1alpha1` custom resources to `app.terraform.io/v1alpha2` according to the updated schema, meaning that v1 of the operator can no longer serve custom resources. Before patching, update your existing custom resources to satisfy the v2 schema requirements. [Learn more](#manifest-schema-migration).
```

I also suggested using the [suggested callout styling for upgrade warnings](https://github.com/hashicorp/engineering-docs/blob/main/writing/style-guide.md#labels)

$ kubectl --namespace ${RELEASE_NAMESPACE} get secret terraformrc -o json | jq '.data.credentials' | tr -d '"' | base64 -d
```

> Update existing custom resources accoring to the migration table above and apply your changes.

Reviewer: ```suggestion
Update existing custom resources [according to the schema migration guidance](#manifest-schema-migration) and apply your changes.
```

```shell-session
$ kubectl apply --filename <UPDATED_V2_WORKSPACE_MANIFEST.yaml>
```

Download Workspace CRD patch B.

```shell-session
$ curl -sO https://raw.githubusercontent.com/hashicorp/terraform-cloud-operator/main/docs/migration/crds/workspaces_patch_b.yaml
```

> View the changes that **patch B** will apply to the workspace CRD.

Reviewer: ```suggestion
View the changes that patch B applies to the workspace CRD.
```

```shell-session
$ kubectl diff --filename workspaces_patch_b.yaml
```

> Patch workspace CRD with **patch B**. This patch adds `.status.runStatus` support that was excluded in patch A.

Reviewer: ```suggestion
Patch workspace CRD with patch B. This patch adds `.status.runStatus` support, which was excluded in patch A.
```

```

The `finalizer` exists to provide greater control over the migration process. Verify the custom resource, and when you’re ready to migrate it, use the `kubectl patch` command to update the `finalizer` value.

```shell-session
$ kubectl patch workspace migration --type=merge --patch '{"metadata": {"finalizers": ["workspace.app.terraform.io/finalizer"]}}'
```

Review the operator logs once more and verify there are no error messages.

```shell-session
$ kubectl logs -f <POD_NAME>
```
The operator reconciles resources during the next sync period. This interval is set by the `operator.syncPeriod` configuration of the operator and defaults to five minutes. 

If you have any migrated `Module` custom resources, apply them now.

```shell-session
$ kubectl apply --filename <MIGRATED_V2_MODULE_MANIFEST.yaml>
```

In v2 of the operator, the `applyMethod` is set to `manual` by default. In this case, a new run in a managed workspace requires manual approval. Run the following command for each `Workspace` resource to change it to `auto` approval.

```shell-session

> kubectl patch workspace <WORKSPACE_NAME> --type=merge --patch '{"spec": {"applyMethod": "auto"}}'

Reviewer 2: ```suggestion
$ kubectl patch workspace <WORKSPACE_NAME> --type=merge --patch '{"spec": {"applyMethod": "auto"}}'
```

```
ERROR	Migration	{"workspace": "default/<WORKSPACE_NAME>", "msg": "spec contains old finalizer finalizer.workspace.app.terraform.io"}
```

The `finalizer` exists to provide greater control over the migration process. Verify the custom resource, and when you’re ready to migrate it, use the `kubectl patch` command to update the `finalizer` value.

```shell-session
$ kubectl patch workspace migration --type=merge --patch '{"metadata": {"finalizers": ["workspace.app.terraform.io/finalizer"]}}'
```

> Review the operator logs once more and verify there are no error messages reported.

Reviewer: ```suggestion
Review the operator logs once more and verify there are no error messages.
```

```shell-session
$ kubectl logs -f <POD_NAME>
```

> The operator will reconcile resources during the next sync period. This interval is set by the `operator.syncPeriod` configuration of the operator and defaults to five minutes. 

Reviewer: ```suggestion
The operator reconciles resources during the next sync period. This interval is set by the `operator.syncPeriod` configuration of the operator and defaults to five minutes. 
```

If you have any migrated `Module` custom resources, apply them now.

```shell-session
$ kubectl apply --filename <MIGRATED_V2_MODULE_MANIFEST.yaml>
```

> In v2 of the operator, `applyMethod` is set to `manual` by default. In this case, a new run in a managed workspace will require manual approval. Run the following command for each Workspace resource to change it to `auto` approval.

Reviewer: ```suggestion
In v2 of the operator, the `applyMethod` is set to `manual` by default. In this case, a new run in a managed workspace requires manual approval. Run the following command for each `Workspace` resource to change it to `auto` approval.
```
